### ISSUE TITLE
List.Nonempty.foldl and foldr aren't the traditional foldl and foldr

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/base_v1#65]
See the following type signatures:

`List.Nonempty.foldl : (a ->{e} a ->{e} a) -> Nonempty a ->{e} a`
`List.Nonempty.foldr : (a ->{e} a ->{e} a) -> Nonempty a ->{e} a`

Contrast these to `foldl` and `foldr` on `List`:
`List.foldl : (b ->{𝕖} a ->{𝕖} b) -> b -> [a] ->{𝕖} b`
`List.foldr : (a ->{e} b ->{e} b) -> b -> [a] ->{e} b`

The `Nonempty` versions don't allow an accumulator type that is different than the list element type. Likewise, they don't take an initial accumulator value. At least in the Scala world, I've seen these referred to as `reduce` instead of `fold`.

This could be a fairly straightforward PR that renames some terms, changes some linked docs, and adds new `fold` versions of these methods. But I got tripped up on a bikeshedding thing: the parallel of `foldl` and `foldr` would be `reducel` and `reducer`, but `reducer` doesn't convey the same meaning that `reduceRight` would. This kind of got me thinking that `foldl` and `foldr` might violate Unison's principle of not needlessly abbreviating terms.

I know that one of the great things about Unison is that it's easy to rename things after the fact, but since the original problem was a term that probably wasn't named what it should have been, it kind of seems like it's time to choose the preferred name here. WDYT?

**Comment by aryairani** [comment]
Even if it's easy to change names in Unison, I think the base lib should be as clear and inviting as possible for newcomers.  I expect a Haskeller looking for `reducer` is not going to have a hard time recognizing `reduceRight`, but the reverse might not be true.

So that would be a "yes" from me re. `foldl`/`foldr` signature symmetry, "yes" to distinguishing `fold*` from `reduce*` (and maybe an extra alias from `reduce` to whichever of `reduceLeft`?), and "yes" to full names, and I'd welcome any further `base` naming principles as well.

**Comment by ceedubs** [comment]
@aryairani thanks for the quick feedback. I plan to give this a try in the upcoming days.

**Comment by philipschwarz** [comment]
FWIW, I think in Haskell the top two signatures are called foldl1 and foldr1

**Comment by philipschwarz** [comment]
e.g. from https://www.cs.nott.ac.uk/~pszgmh/pih.html:
![image](https://user-images.githubusercontent.com/116804/89499983-d2deea00-d7b8-11ea-9294-c5b0a6fa7e2f.png)

**Comment by runarorama** [comment]
Changing these names around should be very easy, and any linked `Doc` values should update automatically. I kind of think `foldLeft` and `foldRight` are better names than `foldl` and `foldr`.

I don't think adding `1` to names clarifies anything, and we should avoid that kind of thing in general.

**Comment by ceedubs** [comment]
@runarorama are you saying that you'd like me to rename existing `foldr` and `foldl` terms as part of this PR? I'm happy to do that (though it might not be today).

**Comment by runarorama** [comment]
Whatever you think is best!

**Comment by ceedubs** [comment]
@runarorama I've updated my PR to rename existing `foldl` and `foldr` terms to `foldLeft` and `foldRight`. Here is the updated output:

```
The changes summarized below are available for you to review, using the
following command:

  pull-request.load git@github.com:unisonweb/base:.trunk git@github.com:ceedubs/unison-dev:.prs._base.nelfoldl

Added definitions:

   List.Nonempty.foldLeft                        : (b ->{e} a ->{e} b)
                                                 -> b
                                                 -> Nonempty a
                                                 ->{e} b (+3 metadata)
   ┌ List.Nonempty.foldLeft.doc                  : Doc (+2 metadata)
   └ List.foldLeft.doc                           : Doc (+2 metadata)
   List.Nonempty.foldLeft.tests.listConsistency  : [Result] (+3 metadata)
   List.Nonempty.foldLeft.tests.multiple         : [Result] (+3 metadata)
   List.Nonempty.foldLeft.tests.single           : [Result] (+3 metadata)
   List.Nonempty.foldRight                       : (a ->{e} b ->{e} b)
                                                 -> b
                                                 -> Nonempty a
                                                 ->{e} b (+2 metadata)
   List.Nonempty.foldRight.tests.listConsistency : [Result] (+3 metadata)
   List.Nonempty.foldRight.tests.multiple        : [Result] (+3 metadata)
   List.Nonempty.foldRight.tests.single          : [Result] (+3 metadata)
   test.gen.nonemptyListOf                       : '{Gen} a
                                                 -> '{Gen} Nonempty a (+2 metadata)

Name changes:

  Original                          Changes
   List.Nonempty.foldl               List.Nonempty.reduceLeft (added)
                                     List.Nonempty.foldl (removed)
  
   List.Nonempty.foldl.doc           List.Nonempty.reduceLeft.doc (added)
                                     List.Nonempty.foldl.doc (removed)
  
   List.Nonempty.foldl.test          List.Nonempty.reduceLeft.test (added)
                                     List.Nonempty.foldl.test (removed)
  
   List.Nonempty.foldr               List.Nonempty.reduceRight (added)
                                     List.Nonempty.foldr (removed)
  
   List.Nonempty.foldr.doc           List.Nonempty.reduceRight.doc (added)
                                     List.Nonempty.foldr.doc (removed)
  
   List.Nonempty.foldr.test          List.Nonempty.reduceRight.test (added)
                                     List.Nonempty.foldr.test (removed)
  
   List.foldl                        List.foldLeft (added)
                                     List.foldl (removed)
  
   List.foldr                        List.foldRight (added)
                                     List.foldr (removed)
  
   List.foldr.doc                    List.foldRight.doc (added)
                                     List.foldr.doc (removed)
  
   List.foldr.tests.homomorphism     List.foldRight.tests.homomorphism (added)
                                     List.foldr.tests.homomorphism (removed)
  
   Set.foldl                         Set.foldLeft (added)
                                     Set.foldl (removed)
  
   Set.foldl.doc                     Set.foldLeft.doc (added)
                                     Set.foldl.doc (removed)
  
   Set.foldl.tests.homomorphism      Set.foldLeft.tests.homomorphism (added)
                                     Set.foldl.tests.homomorphism (removed)
  
   Set.foldr                         Set.foldRight (added)
                                     Set.foldr (removed)
  
   Set.foldr.doc                     Set.foldRight.doc (added)
                                     Set.foldr.doc (removed)
  
   Set.foldr.tests.homomorphism      Set.foldRight.tests.homomorphism (added)
                                     Set.foldr.tests.homomorphism (removed)

```

**Comment by ceedubs** [comment]
Oops in my last couple of comments I was mixing up this "issue" issue with the "PR" issue #67.

**Comment by runarorama** [comment]
Fixed in #67
